contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaPartitionScanSpecBuilder.java

```diff
@@ -0,0 +1,346 @@
+
+    /*
+    We should handle the case where the specified offset does not exist in the current context,
+    i.e., fieldValue < startOffset or fieldValue > endOffset in a particular topic-partition.
+    Else, KafkaConsumer.poll will throw "TimeoutException".
+    */
+
+    switch (functionName) {
+      case "equal":
+        for(TopicPartition tp : topicPartitions) {
+          if(fieldValue < fullScanSpec.get(tp).getStartOffset()) {
+            //Offset does not exist
+            scanSpec.add(
+                new KafkaPartitionScanSpec(tp.topic(), tp.partition(),
+                    fullScanSpec.get(tp).getEndOffset(), fullScanSpec.get(tp).getEndOffset()));
+          } else {
+            long val = Math.min(fieldValue, fullScanSpec.get(tp).getEndOffset());
+            long nextVal = Math.min(val+1, fullScanSpec.get(tp).getEndOffset());
+            scanSpec.add(new KafkaPartitionScanSpec(tp.topic(), tp.partition(), val, nextVal));
+          }
+        }
+        break;
+      case "greater_than_or_equal_to":
+        for(TopicPartition tp : topicPartitions) {
+          //Ensure scan range is between startOffset and endOffset,
+          long val = Math.max(fullScanSpec.get(tp).getStartOffset(),
+              Math.min(fieldValue, fullScanSpec.get(tp).getEndOffset()));
+          scanSpec.add(
+              new KafkaPartitionScanSpec(tp.topic(), tp.partition(), val,
+                  fullScanSpec.get(tp).getEndOffset()));
+        }
+        break;
+      case "greater_than":
+        for(TopicPartition tp : topicPartitions) {
+          //Ensure scan range is between startOffset and endOffset,
+          long val = Math.max(fullScanSpec.get(tp).getStartOffset(),
+              Math.min(fieldValue+1, fullScanSpec.get(tp).getEndOffset()));
+
+          scanSpec.add(
+              new KafkaPartitionScanSpec(tp.topic(), tp.partition(),
+                  val, fullScanSpec.get(tp).getEndOffset()));
+        }
+        break;
+      case "less_than_or_equal_to":
+        for(TopicPartition tp : topicPartitions) {
+          //Ensure scan range is between startOffset and endOffset,
+          long val = Math.max(fullScanSpec.get(tp).getStartOffset(),
+              Math.min(fieldValue+1, fullScanSpec.get(tp).getEndOffset()));
+
+          scanSpec.add(
+              new KafkaPartitionScanSpec(tp.topic(), tp.partition(),
+                  fullScanSpec.get(tp).getStartOffset(), val));
+        }
+        break;
+      case "less_than":
+        for(TopicPartition tp : topicPartitions) {
+          //Ensure scan range is between startOffset and endOffset,
+          long val = Math.max(fullScanSpec.get(tp).getStartOffset(),
+              Math.min(fieldValue, fullScanSpec.get(tp).getEndOffset()));
+
+          scanSpec.add(
+              new KafkaPartitionScanSpec(tp.topic(), tp.partition(),
+                  fullScanSpec.get(tp).getStartOffset(), val));
+        }
+        break;
+    }
```

Comment:
Can you please refactor the above code?

Reply:
@aravi5  This is a minor comment for readability. All the above cases can be extracted into one method by taking filedValue or fieldValue+1 as arguments.

Reply:
Got it! I have extracted that to a small utility method.